runtime/buffer.h

```diff
@@ -334,51 +334,41 @@ inline bool can_fuse(const dim& inner, const dim& outer) {
   return inner.stride() * inner.extent() == outer.stride();
 }
 
-template <typename F>
-void for_each_contiguous_slice(void* base, const dim* dims, int d, index_t elem_size, const F& f,
-    index_t slice_extent = 1, index_t extent = 1) {
-  if (d == -1) {
-    // We've handled all the loops, call the function.
-    f(base, slice_extent);
-    return;
-  }
-
-  const slinky::dim& dim = dims[d];
-  extent *= dim.extent();
-  if (extent <= 0) {
-    // Don't want to worry about empty dimensions in the cases below.
-    return;
-  }
+struct for_each_contiguous_slice_dim {
+  union {
```

Comment:
style nit: IMHO would be better to give this a name (e.g. `u`) and then access the fields through it `eg `next->u.stride`) to make it more obvious at the point of read/write that this is a union field.

Reply:
The way I like to write stuff like this is that the read/write shouldn't care if its a union or not. Added comments on each impl type to clarify.